### PR TITLE
Refactor controlplane into a go plugin

### PR DESCRIPTION
**Pull request by mattstep**
Wait to merge this until I add some docs for it. Just getting this up to start a review now.

### CODE REVIEW

File: poolmanager/server/controlplane/controlplane.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
+/*
+	Interface between the Node Pool Manager and the Control Plane
+*/
+
+package controlplane
+
+const CAPACITY_PER_RUNNER = 4096
```
Comment (rdallman) [inline poolmanager/server/controlplane/controlplane.go]: Go's idiom for constants is mixed casing for constants https://github.com/golang/go/wiki/CodeReviewComments#mixed-caps (and they note this is odd, but the stdlib is this way)

File: poolmanager/server/controlplane/controlplane.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+	Address string
+	// Other: certs etc here as managed and installed by CP
+	Capacity int64
+}
+
+type ControlPlane interface {
+	GetLBGRunners(lgbId string) ([]*Runner, error)
+	ProvisionRunners(lgbId string, n int) (int, error)
+	RemoveRunner(lbgId string, id string) error
+}
```
Comment (rdallman) [inline poolmanager/server/controlplane/controlplane.go]: 🥇

File: poolmanager/server/controlplane/plugin/noop.go [hunk 3]
```diff
@@ -1,56 +1,48 @@
...
+	"github.com/fnproject/fn/poolmanager/server/controlplane"
+)
 
-type ControlPlane interface {
-	GetLBGRunners(lgbId string) ([]*Runner, error)
-	ProvisionRunners(lgbId string, n int) (int, error)
-	RemoveRunner(lbgId string, id string) error
-}
+const (
+	EnvFixedRunners = "FN_RUNNER_ADDRESSES"
```
Comment (rdallman) [inline poolmanager/server/controlplane/plugin/noop.go]: ditto

File: poolmanager/server/controlplane/plugin/noop.go [hunk 4]
```diff
@@ -121,3 +113,13 @@ func (cp *controlPlane) RemoveRunner(lbgId string, id string) error {
...
+
+func getEnv(key string) string {
+	value, ok := os.LookupEnv(key)
+	if !ok {
+		log.Panicf("Missing config key: %v", key)
+	}
+	return value
+}
+
+var ControlPlane noopControlPlane
```
Comment (rdallman) [inline poolmanager/server/controlplane/plugin/noop.go]: not sure about this one, it has uninitialized maps and `NewControlPlane()` returns an instance of one of these to get a handle on, perhaps this was cruft from shifting stuff around?

File: poolmanager/server/controlplane/plugin/noop.go [hunk 5]
```diff
@@ -1,56 +1,48 @@
-/*
-	Interface between the Node Pool Manager and the Control Plane
-*/
-
-package cp
+/**
+ * Dummy implementation for the controlplane that just adds delays
+ */
+package main
```
Comment (rdallman) [inline poolmanager/server/controlplane/plugin/noop.go]: package noop? package noopcp? I do not see a main in this package checked in, at least.

**Comment by rdallman** [comment]
thanks for this, looks great. some minor comments

File: poolmanager/server/main.go [hunk 6]
```diff
@@ -116,6 +115,25 @@ func createGrpcCreds(cert string, key string, ca string) (grpc.ServerOption, err
 	return grpc.Creds(creds), nil
 }
 
+func newPluggableControlPlane() controlplane.ControlPlane {
```
Comment (faiq) [inline poolmanager/server/main.go]: Do we know for sure that this instantionalizes the object for sure? I think @rdallman also commented on the bottom of the NoopControlPlane with a similar concern.

We could return a factory as an export, and call the function that was exported.

File: poolmanager/manager.go [hunk 7]
```diff
@@ -8,7 +8,7 @@ import (
 	"sync"
 
 	model "github.com/fnproject/fn/poolmanager/grpc"
-	"github.com/fnproject/fn/poolmanager/server/cp"
```
Comment (faiq) [inline poolmanager/manager.go]: HOW WAS EVERYTHING NAMED CP 👍  on the change

File: poolmanager/manager.go [hunk 8]
```diff
@@ -296,20 +296,20 @@ func (lbg *lbGroup) target(ts time.Time, target int64) {
...
 				// Some kind of error during attempt to scale up
 				logrus.WithError(err).Error("Error occured during attempt to scale up")
 				return
 			}
-			lbg.target_capacity += int64(asked_for) * cp.CAPACITY_PER_RUNNER
+			lbg.target_capacity += int64(asked_for) * controlplane.CAPACITY_PER_RUNNER
 		}
 
-	} else if desiredScale <= lbg.current_capacity-cp.CAPACITY_PER_RUNNER {
+	} else if desiredScale <= lbg.current_capacity-controlplane.CAPACITY_PER_RUNNER {
```
Comment (faiq) [inline poolmanager/manager.go]: can-we_do_someTHING-About-THIS_too?

**Comment by faiq** [comment]
Looks good! Love the plugin stuff. My suggestion would be to have a factory be exposed, and then call the returning function to get an instance of a control plane manager. Right now I'm under the impression that you're not returning an instance of the object with all the maps and stuff allocated.

I think an easy test would be adding another method to the interface that is something like `Name()` that can print the name. This can make it clearer which driver we're using and can ensure that the object is there.

File: poolmanager/server/controlplane/plugin/noop.go (hunk 4 above)
Comment (skinowski) [reply poolmanager/server/controlplane/plugin/noop.go]: probably needs to be initialized in init() which plugin.Open() would call, right?

File: poolmanager/server/controlplane/plugin/noop.go (hunk 5 above)
Comment (mattstep) [reply poolmanager/server/controlplane/plugin/noop.go]: It's a requirement of buildmode=plugin, the symbols are only exported from the main package

File: poolmanager/manager.go (hunk 8 above)
Comment (mattstep) [reply poolmanager/manager.go]: I'll do a followup pr to clean all this up

File: poolmanager/server/controlplane/plugin/noop.go (hunk 4 above)
Comment (mattstep) [reply poolmanager/server/controlplane/plugin/noop.go]: fixed

File: poolmanager/server/main.go (hunk 6 above)
Comment (mattstep) [reply poolmanager/server/main.go]: yep,fixed

**Comment by mattstep** [comment]
@rdallman ptal

File: poolmanager/server/controlplane/plugin/noop.go (hunk 3 above)
Comment (rdallman) [reply poolmanager/server/controlplane/plugin/noop.go]: d'oh

**Pull request merged**